File: problems/9-spiral-text/solution.py
def print_spiral(s):
    import math

    # Calculate the size of the matrix
    n = math.ceil(math.sqrt(len(s)))

    # Initialize the matrix
    matrix = [[" "] * n for _ in range(n)]

    # Calculate the boundaries
    top = 0
    bottom = n - 1
    left = 0
    right = n - 1

    # Initialize the string index
    index = 0

    while True:
        # Print the top row
        for i in range(left, right + 1):
            if index >= len(s):
                break

            matrix[top][i] = s[index]
            index += 1
        top += 1
        if not (top <= bottom):
            break

        # Print the right column
        for i in range(top, bottom + 1):
            if index >= len(s):
                break

            matrix[i][right] = s[index]
            index += 1
        right -= 1
        if not (left <= right):
            break

        # Print the bottom row
        for i in range(right, left - 1, -1):
            if index >= len(s):
                break

            matrix[bottom][i] = s[index]
            index += 1
        bottom -= 1
        if not (top <= bottom):
            break

        # Print the left column
        for i in range(bottom, top - 1, -1):
            if index >= len(s):
                break

            matrix[i][left] = s[index]
            index += 1
        left += 1
        if not (left <= right):
            break

        if index >= len(s):
            break

    # Print the matrix
    for row in matrix:
        print(" ".join(row))


s = input()
print_spiral(s)

File: problems/9-spiral-text/solution.java
import java.lang.Math;
import java.util.Scanner;
import java.util.Arrays;

public class solution {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String s = scanner.nextLine();
        printSpiral(s);
    }

    public static void printSpiral(String s) {
        int n = (int) Math.ceil(Math.sqrt(s.length()));

        // Initialize the matrix
        char[][] matrix = new char[n][n];
        for (char[] row: matrix) {
            Arrays.fill(row, ' ');
        }

        // Calculate the boundaries
        int top = 0;
        int bottom = n - 1;
        int left = 0;
        int right = n - 1;

        // Initialize the string index
        int index = 0;

        while (true) {
            // Print the top row
            for (int i = left; i <= right; i++) {
                if (index >= s.length()) {
                    break;
                }
                matrix[top][i] = s.charAt(index);
                index++;
            }
            top++;
            if (!(top <= bottom)) {
                break;
            }

            // Print the right column
            for (int i = top; i <= bottom; i++) {
                if (index >= s.length()) {
                    break;
                }
                matrix[i][right] = s.charAt(index);
                index++;
            }
            right--;
            if (!(left <= right)) {
                break;
            }

            // Print the bottom row
            for (int i = right; i >= left; i--) {
                if (index >= s.length()) {
                    break;
                }
                matrix[bottom][i] = s.charAt(index);
                index++;
            }
            bottom--;
            if (!(top <= bottom)) {
                break;
            }

            // Print the left column
            for (int i = bottom; i >= top; i--) {
                if (index >= s.length()) {
                    break;
                }
                matrix[i][left] = s.charAt(index);
                index++;
            }
            left++;
            if (!(left <= right)) {
                break;
            }

            if (index >= s.length()) {
                break;
            }
        }

        // Print the matrix
        for (char[] row : matrix) {
            boolean first = true;
            for (char c : row) {
                if (first) {
                    first = false;
                } else {
                    System.out.print(" ");
                }
                System.out.print(c);
            }
            System.out.println();
        }
    }
}